[quant][graphmode] Integration tests for initial quantization graph mode

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25127 [quant][graphmode] insert_quant_dequant work with qconfig_dict
* #25069 [graphmode] insert_observers use qconfig_dict
* #25000 [cleanup] Remove InsertQuantDeQuantNode
* #24999 [cleanup] Remove insert_observers pass
* #24998 [cleanup] Remove deprecated graph mode quantization tests
* **#24428 [quant][graphmode] Integration tests for initial quantization graph mode**
* #24427 [quant][graphmode] quant_fusion jit pass
* #24426 [quant][graphmode] insert_quant_dequant jit pass

Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17001136](https://our.internmc.facebook.com/intern/diff/D17001136)

jerryzh168 committed:
[quant][graphmode] Integration tests for initial quantization graph mode

Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -222,3 +225,110 @@ def forward(self, x):
+
+        ScriptedObserver = torch.jit.script(Observer())
+        ScriptedWeightObserver = torch.jit.script(WeightObserver())
+        print('--------- 1. Prepare Quant --------')
```
I think we shouldn't have print statements in our tests. They'll clobber logs for no reason.

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -222,3 +225,110 @@ def forward(self, x):
+        script_module = TestScriptM()
+        script_module.conv.weight = torch.nn.Parameter(eager_module.conv.weight.detach())
+        quantized_eager_module = quantize(eager_module, default_eval_fn, data)
+        torch._C._jit_set_inline_everything_mode(False)
```
It's better to use the corresponding decorator on a function to make sure we don't affect other tests.

jerryzh168 replied:
we actually need to inline for now and don't inline will be supported later, this line has no effect right now since it is after creation of ScriptModule

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -222,3 +225,110 @@ def forward(self, x):
+
+        def get_forward(m):
+            return m._c._get_method('forward')
+        # Script mode
```
Nit: "# Script mode" comment should be a couple lines earlier.

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -222,3 +225,110 @@ def forward(self, x):
+        def get_forward(m):
+            return m._c._get_method('forward')
+        # Script mode
+        # TODO: test jit.script as well
```
What does this TODO mean?

jerryzh168 replied:
here we define a ScriptModule, but we want to test the path when we get a module from `torch.jit.script` as well, I'll add a test in qconfig prs.

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -222,3 +225,110 @@ def forward(self, x):
+        script_module._c = torch._C._jit_pass_insert_quant_dequant(script_module._c, "forward")
+        # Note that observer modules are not removed right now
+        # print(script_module._c._get_modules())
+        # torch._C._jit_pass_custom_pattern_based_rewrite_graph()
```
Looks like a leftover.

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -222,3 +225,110 @@ def forward(self, x):
+        self.assertEqual(eager_result, script_result)
+        # Compare results for eager and graph mode
+
+    def test_module_rewriter(self):
```
This test seems to do nothing.

ZolotukhinM commented on test/test_quantizer.py:
```diff
@@ -222,3 +225,110 @@ def forward(self, x):
         self.assertAlmostEqual(eagerDict["z"][0], activationDict["z.1"][0], places=15)
         self.assertAlmostEqual(eagerDict["z"][1], activationDict["z.1"][1], places=15)
+
+    def test_compare_qparam_eager_script_default(self):
```
Nit: what do 'qparam' and 'default' words mean in the name of this test?

jerryzh168 replied:
I plan to rename this to test_default in later PR..
default just means simple/naive situation, maybe I can rename that to test_simple

ZolotukhinM reviewed (approved):
We have the end-to-end flow working, sweet!

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17001136](https://our.internmc.facebook.com/intern/diff/D17001136)

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17001136](https://our.internmc.facebook.com/intern/diff/D17001136)

jerryzh168 committed:
Update on "[quant][graphmode] Integration tests for initial quantization graph mode"


Summary:
att

Test Plan:
python test/test_quantizer.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17001136](https://our.internmc.facebook.com/intern/diff/D17001136)
